File: racer.go
// racer.go - Гоночная игра в консоли (Go)
package main

import (
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"os/exec"
	"runtime"
	"time"

	"github.com/eiannone/keyboard"
)

const (
	WIDTH      = 11
	HEIGHT     = 15
	ROAD_LEFT  = 2
	ROAD_RIGHT = WIDTH - 3
	CAR        = '▲'
	OBSTACLE   = '█'
	RECORD_FILE = "racer_record.json"
)

type Racer struct {
	playerPos int
	obstacles [][2]int // [x, y]
	score     int
	speed     int
	gameOver  bool
	paused    bool
	running   bool
	record    int
}

func NewRacer() *Racer {
	return &Racer{
		playerPos: WIDTH / 2,
		obstacles: make([][2]int, 0),
		score:     0,
		speed:     1,
		gameOver:  false,
		paused:    false,
		running:   true,
		record:    loadRecord(),
	}
}

func loadRecord() int {
	file, err := os.Open(RECORD_FILE)
	if err != nil {
		return 0
	}
	defer file.Close()
	var data map[string]int
	decoder := json.NewDecoder(file)
	if err := decoder.Decode(&data); err != nil {
		return 0
	}
	if val, ok := data["record"]; ok {
		return val
	}
	return 0
}

func saveRecord(record int) {
	data := map[string]int{"record": record}
	file, _ := os.Create(RECORD_FILE)
	defer file.Close()
	encoder := json.NewEncoder(file)
	encoder.Encode(data)
}

func (r *Racer) spawnObstacle() {
	x := ROAD_LEFT + rand.Intn(ROAD_RIGHT-ROAD_LEFT+1)
	r.obstacles = append(r.obstacles, [2]int{x, 0})
}

func (r *Racer) update() {
	if r.gameOver || r.paused {
		return
	}

	// Сдвиг вниз
	newObs := make([][2]int, 0, len(r.obstacles))
	for _, obs := range r.obstacles {
		obs[1]++
		if obs[1] < HEIGHT {
			newObs = append(newObs, obs)
		}
	}
	r.obstacles = newObs

	// Столкновение
	for _, obs := range r.obstacles {
		if obs[1] == HEIGHT-1 && obs[0] == r.playerPos {
			r.gameOver = true
			if r.score > r.record {
				r.record = r.score
				saveRecord(r.record)
			}
			return
		}
	}
	r.score++
	r.speed = 1 + r.score/10

	// Спавн
	prob := 0.3 + min(0.4, float64(r.score)/200.0)
	if rand.Float64() < prob {
		r.spawnObstacle()
	}
}

func (r *Racer) draw() {
	clearScreen()
	fmt.Println(stringRepeat("═", WIDTH))
	fmt.Printf("  Счёт: %d   Рекорд: %d   Скорость: %d\n", r.score, r.record, r.speed)
	fmt.Println(stringRepeat("═", WIDTH))

	for y := 0; y < HEIGHT; y++ {
		line := make([]rune, WIDTH)
		for x := 0; x < WIDTH; x++ {
			if x == 0 || x == WIDTH-1 {
				line[x] = '│'
			} else {
				isObs := false
				for _, obs := range r.obstacles {
					if obs[0] == x && obs[1] == y {
						isObs = true
						break
					}
				}
				if isObs {
					line[x] = OBSTACLE
				} else if y == HEIGHT-1 && x == r.playerPos {
					line[x] = CAR
				} else {
					line[x] = ' '
				}
			}
		}
		fmt.Println(string(line))
	}
	fmt.Println(stringRepeat("═", WIDTH))
	status := "ИГРА"
	if r.paused {
		status = "ПАУЗА"
	}
	fmt.Printf("  %s  |  ← → движение  |  P - пауза  |  R - рестарт  |  Q - выход\n", status)
}

func (r *Racer) reset() {
	r.playerPos = WIDTH / 2
	r.obstacles = nil
	r.score = 0
	r.speed = 1
	r.gameOver = false
	r.paused = false
}

func (r *Racer) handleInput() {
	for r.running {
		char, key, err := keyboard.GetKey()
		if err != nil {
			continue
		}
		switch key {
		case keyboard.KeyArrowLeft:
			if !r.gameOver && !r.paused {
				if r.playerPos > ROAD_LEFT {
					r.playerPos--
				}
			}
		case keyboard.KeyArrowRight:
			if !r.gameOver && !r.paused {
				if r.playerPos < ROAD_RIGHT {
					r.playerPos++
				}
			}
		case keyboard.KeyPgUp, keyboard.KeyPgDn: // ignore
		default:
			switch char {
			case 'a', 'A':
				if !r.gameOver && !r.paused {
					if r.playerPos > ROAD_LEFT {
						r.playerPos--
					}
				}
			case 'd', 'D':
				if !r.gameOver && !r.paused {
					if r.playerPos < ROAD_RIGHT {
						r.playerPos++
					}
				}
			case 'p', 'P':
				r.paused = !r.paused
			case 'r', 'R':
				r.reset()
			case 'q', 'Q':
				r.running = false
				return
			}
		}
	}
}

func clearScreen() {
	cmd := exec.Command("clear")
	if runtime.GOOS == "windows" {
		cmd = exec.Command("cmd", "/c", "cls")
	}
	cmd.Stdout = os.Stdout
	cmd.Run()
}

func stringRepeat(s string, n int) string {
	res := ""
	for i := 0; i < n; i++ {
		res += s
	}
	return res
}

func min(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}

func (r *Racer) run() {
	go r.handleInput()

	ticker := time.NewTicker(time.Second / 10) // 10 fps
	defer ticker.Stop()

	lastUpdate := time.Now()
	for r.running {
		select {
		case <-ticker.C:
			now := time.Now()
			if now.Sub(lastUpdate) >= time.Second/time.Duration(r.speed) {
				r.update()
				r.draw()
				lastUpdate = now
			}
		}
	}
}

func main() {
	rand.Seed(time.Now().UnixNano())
	racer := NewRacer()
	racer.run()
	fmt.Println("Игра завершена.")
}

File: racer.rs
// racer.rs - Гоночная игра в консоли (Rust)
use crossterm::{
    event::{self, Event, KeyCode},
    execute,
    terminal::{self, Clear, ClearType},
};
use std::io::{stdout, Write};
use std::time::{Duration, Instant};
use std::collections::VecDeque;
use serde::{Serialize, Deserialize};
use std::fs;
use rand::Rng;

const WIDTH: usize = 11;
const HEIGHT: usize = 15;
const ROAD_LEFT: usize = 2;
const ROAD_RIGHT: usize = WIDTH - 3;
const CAR: char = '▲';
const OBSTACLE: char = '█';
const RECORD_FILE: &str = "racer_record.json";

#[derive(Serialize, Deserialize)]
struct RecordData {
    record: u32,
}

struct Racer {
    player_pos: usize,
    obstacles: VecDeque<(usize, usize)>,
    score: u32,
    speed: u32,
    game_over: bool,
    paused: bool,
    running: bool,
    record: u32,
    rng: rand::rngs::ThreadRng,
}

impl Racer {
    fn new() -> Self {
        Self {
            player_pos: WIDTH / 2,
            obstacles: VecDeque::new(),
            score: 0,
            speed: 1,
            game_over: false,
            paused: false,
            running: true,
            record: Self::load_record(),
            rng: rand::thread_rng(),
        }
    }

    fn load_record() -> u32 {
        if let Ok(data) = fs::read_to_string(RECORD_FILE) {
            if let Ok(rec) = serde_json::from_str::<RecordData>(&data) {
                return rec.record;
            }
        }
        0
    }

    fn save_record(record: u32) {
        let data = RecordData { record };
        let _ = fs::write(RECORD_FILE, serde_json::to_string(&data).unwrap());
    }

    fn spawn_obstacle(&mut self) {
        let x = self.rng.gen_range(ROAD_LEFT..=ROAD_RIGHT);
        self.obstacles.push_back((x, 0));
    }

    fn update(&mut self) {
        if self.game_over || self.paused {
            return;
        }

        // Сдвиг вниз
        for item in self.obstacles.iter_mut() {
            item.1 += 1;
        }
        // Удалить вышедшие
        self.obstacles.retain(|&(_, y)| y < HEIGHT);

        // Столкновение
        for &(x, y) in &self.obstacles {
            if y == HEIGHT - 1 && x == self.player_pos {
                self.game_over = true;
                if self.score > self.record {
                    self.record = self.score;
                    Self::save_record(self.record);
                }
                return;
            }
        }
        self.score += 1;
        self.speed = 1 + self.score / 10;

        // Спавн
        let prob = 0.3 + f64::min(0.4, self.score as f64 / 200.0);
        if self.rng.gen_bool(prob) {
            self.spawn_obstacle();
        }
    }

    fn draw(&self) {
        execute!(stdout(), Clear(ClearType::All)).unwrap();
        let mut out = stdout();
        writeln!(out, "{}", "═".repeat(WIDTH)).unwrap();
        writeln!(out, "  Счёт: {}   Рекорд: {}   Скорость: {}", self.score, self.record, self.speed).unwrap();
        writeln!(out, "{}", "═".repeat(WIDTH)).unwrap();

        for y in 0..HEIGHT {
            let mut line = String::with_capacity(WIDTH);
            for x in 0..WIDTH {
                if x == 0 || x == WIDTH - 1 {
                    line.push('│');
                } else {
                    let is_obs = self.obstacles.iter().any(|&(ox, oy)| ox == x && oy == y);
                    if is_obs {
                        line.push(OBSTACLE);
                    } else if y == HEIGHT - 1 && x == self.player_pos {
                        line.push(CAR);
                    } else {
                        line.push(' ');
                    }
                }
            }
            writeln!(out, "{}", line).unwrap();
        }
        writeln!(out, "{}", "═".repeat(WIDTH)).unwrap();
        let status = if self.paused { "ПАУЗА" } else { "ИГРА" };
        writeln!(out, "  {}  |  ← → движение  |  P - пауза  |  R - рестарт  |  Q - выход", status).unwrap();
        out.flush().unwrap();
    }

    fn reset(&mut self) {
        self.player_pos = WIDTH / 2;
        self.obstacles.clear();
        self.score = 0;
        self.speed = 1;
        self.game_over = false;
        self.paused = false;
    }

    fn run(&mut self) {
        // Настройка терминала для raw режима
        terminal::enable_raw_mode().unwrap();
        let mut last_update = Instant::now();

        while self.running {
            // Обработка событий клавиатуры
            if event::poll(Duration::from_millis(50)).unwrap() {
                if let Event::Key(key) = event::read().unwrap() {
                    match key.code {
                        KeyCode::Left | KeyCode::Char('a') | KeyCode::Char('A') => {
                            if !self.game_over && !self.paused && self.player_pos > ROAD_LEFT {
                                self.player_pos -= 1;
                            }
                        }
                        KeyCode::Right | KeyCode::Char('d') | KeyCode::Char('D') => {
                            if !self.game_over && !self.paused && self.player_pos < ROAD_RIGHT {
                                self.player_pos += 1;
                            }
                        }
                        KeyCode::Char('p') | KeyCode::Char('P') => {
                            self.paused = !self.paused;
                        }
                        KeyCode::Char('r') | KeyCode::Char('R') => {
                            self.reset();
                        }
                        KeyCode::Char('q') | KeyCode::Char('Q') => {
                            self.running = false;
                        }
                        _ => {}
                    }
                }
            }

            // Обновление по времени
            let now = Instant::now();
            if now - last_update >= Duration::from_secs_f64(1.0 / self.speed as f64) {
                self.update();
                self.draw();
                last_update = now;
            }
        }

        terminal::disable_raw_mode().unwrap();
    }
}

fn main() {
    let mut racer = Racer::new();
    racer.run();
    println!("Игра завершена.");
}
